### ISSUE TITLE
Core Data: optional scalar attributes

### ISSUE THREAD

**Issue by djbe** [SwiftGen/SwiftGen#565]
Currently we ignore optionality for scalar attributes, which isn't ideal.

To add support for this, and to do it correctly, we'll need something like this for each optional scalar attribute (based on mogenerator):
```swift
public var counter: Int16? {
  get {
    let key = "counter"
    willAccessValue(forKey: key)
    defer { didAccessValue(forKey: key) }

    return primitiveValue(forKey: key) as? Int16
  }
  set {
    let key = "counter"
    willChangeValue(forKey: key)
    defer { didChangeValue(forKey: key) }

    setPrimitiveValue(newValue, forKey: key)
  }
}
```

Note that:

- We can't use `#keyPath(counter)` because it isn't an obj-c property.
- We can't make it an `@objc` property because `Int16?` can't be represented in obj-c.
- We can't use Swift key paths, because we can't access the string representation (yet): https://bugs.swift.org/browse/SR-5220

Reference mogenerator template: https://github.com/rentzsch/mogenerator/blob/swift42/templates/machine.swift.motemplate
Article that explains why we can't just use `@NSManaged`:
http://aplus.rs/2017/handling-core-data-optional-scalar-attributes/

**Comment by andreyslyusar** [comment]
@djbe What do you think about using next approach instead:

```
import CoreData

@objc(ExampleManagedObject)
public class ExampleManagedObject: NSManagedObject {
  @NSManaged private var primitiveCounter: NSNumber?
  
  public var counter: Int? {
    set {
      willChangeValue(forKey: "counter")
      defer { didChangeValue(forKey: "counter") }
     
      guard let value = newValue else {
        primitiveCounter = nil
        return
      }
      primitiveCounter = NSNumber(value: value)
    }
    get {
      willAccessValue(forKey: "counter")
      defer { didAccessValue(forKey: "counter") }
    
      return primitiveCounter?.intValue
    }
  }
}
```

- `@NSManaged private var primitiveCounter: NSNumber?`private representation to reference model property `counter`, constructed by `primitive` + `<variable model name>`
- `public var counter: Int?` - public property visible outside of `ExampleManagedObject`. We are using `Int?` instead `Int16` `Int32` or `Int64`, as it will make working with properties easies  in Swift code.
- `set` and `get` method do internal magic to save and extract value, also type casting.

<img width="1411" alt="screenshot 2019-01-14 at 20 03 53" src="https://user-images.githubusercontent.com/1401690/51135327-06a29980-183a-11e9-9448-3dfbb48ab875.png">

**Comment by djbe** [comment]
Hmmm, never liked that alternate name approach. Mogenerator does something similar with `counterValue`.

I'm also not sure if your implementation would work? You're not actually setting/getting the value for the `counter` key. You tell CD that you will, but you don't actually do it. You use the `primitiveCounter` key, which doesn't exist in your model. It also creates conflicts if the user also has an attribute with the same name, which I'd like to avoid at all costs.

Using `Int` instead of `Int16`, ... is another discussion. Not sure if that's a good idea either, as it hides that detail from the user, who won't know that larger values will overflow and may cause undefined behaviour/crashes.

**Comment by MauriceArikoglu** [comment]
+1 on this

**Comment by andreyslyusar** [comment]
> I'm also not sure if your implementation would work? You're not actually setting/getting the value for the counter key. You tell CD that you will, but you don't actually do it. You use the primitiveCounter key, which doesn't exist in your model. It also creates conflicts if the user also has an attribute with the same name, which I'd like to avoid at all costs.

I am 100% sure it will work. I can create a `hello world` example with simple save and fetch method. 

> Using Int instead of Int16, ... is another discussion. Not sure if that's a good idea either, as it hides that detail from the user, who won't know that larger values will overflow and may cause undefined behaviour/crashes.

In 90% cases developers will just cast `Int16` to `Int` without overflow check  `Int(int16variable)`.

Do you think we can implement both solutions and make selection based on template generation parameter?

**Comment by djbe** [comment]
Implemented by #567, released in SwiftGen 6.1.0 🎉
